h/static/scripts/controllers/autosuggest-dropdown-controller.js

```diff
@@ -0,0 +1,454 @@
+    this._header.classList.add(this.options.classNames.header);
+    this._setHeader(this.state.header);
+    this._suggestionContainer.appendChild(this._header);
+
+    this._listContainer = document.createElement('ul');
+    this._listContainer.classList.add(this.options.classNames.list);
+    this._suggestionContainer.appendChild(this._listContainer);
+
+    // put the suggestions adjacent to the input element
+    // firefox does not support insertAdjacentElement
+    if(HTMLElement.prototype.insertAdjacentElement){
+      this._input.insertAdjacentElement('afterend', this._suggestionContainer);
+    }else {
+      this._input.parentNode.insertBefore(this._suggestionContainer, this._input.nextSibling);
+    }
+  }
+
+  /**
+   * _renderListItems - updates the content of the list container and builds
+   *  the new set of list items.
+   */
+  _renderListItems(){
+    // Create the new list items, render their contents
+    // and update the dom with the new elements.
+
+    const currentActive = this._getActiveListItemElement();
+    const currentActiveSelectorId = currentActive ? currentActive.getAttribute('data-suggestion-id') : null;
+
+    this._listContainer.innerHTML = '';
+
+    this.state.list.forEach((listItem)=>{
+      let li = document.createElement('li');
+      li.classList.add(this.options.classNames.item);
+      li.setAttribute('data-suggestion-id', listItem.__suggestionId);
+
+      // persist the active state of selections
+      // through the rendering process
+      if(listItem.__suggestionId === currentActiveSelectorId){
+        li.classList.add(this.options.classNames.activeItem);
+      }
+
+      // this should use some sort of event delegation if
+      // we find we want to expand this to lists with *a lot* of items in it
+      // But for now this binding has no real affect on small list perf
+      li.addEventListener('mouseenter', this._toggleItemHoverState.bind(this, /*hovering*/true));
+      li.addEventListener('mouseleave', this._toggleItemHoverState.bind(this, /*hovering*/false));
+      li.addEventListener('mousedown', (event)=>{
+        // for situations like mobile, hovering might not be
+        // the first event to set the active state for an element
+        // so we will mimic that on mouse down and let selection happen
+        // at the top level event
+        this._toggleItemHoverState(/*hovering*/true, event);
+        this._selectCurrentActiveItem();
+      });
+
+      li.innerHTML = this.options.renderListItem(listItem);
+
+      this._listContainer.appendChild(li);
+    });
+  }
+
+  /**
+   * _addTopLevelEventListeners - The events that can be set on a "global" or top
+   *  level scope, we are going to set them here.
+   */
+  _addTopLevelEventListeners(){
+
+    // we need to use mousedown instead of click
+    // so we can beat the blur event which can
+    // change visibility/target of the active event
+    document.addEventListener('mousedown', (event)=>{
```

Comment:
It's not an upgrade controller so that beforeRemove isn't applicable. It's also not really needed at this stage.